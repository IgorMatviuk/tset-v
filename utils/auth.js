export default function auth(context) {
  return {
    async login() {
      const { email, password } = context.formData
      await context.$auth.loginWith('local', { data: { email, password } })
    },
    async forgotPassword() {
      await context.$axios.$post(
        '/api/customer/forgot-password',
        context.formData
      )
    },
    async register() {
      await context.$axios.$post('/api/customer/register', context.formData)
    }
  }
}
