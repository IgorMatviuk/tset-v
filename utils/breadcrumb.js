export default ($t, $i18n, items) => {
  return [{ text: $t('pages.home'), href: $i18n.path('') }, ...items]
}
