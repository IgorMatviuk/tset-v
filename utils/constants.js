export const CATALOG_ITEMS_PER_PAGE = 24

export const PRODUCT_MAX_QTY = 99

export const DELIVERY_METHODS = {
  NOVAPOSHTA: 'novaposhta',
  LOCAL: 'local',
  ABROAD: 'abroad'
}

export const PAYMENT_METHODS = {
  COD: 'cod',
  PRIVAT: 'privat24',
  LIQPAY: 'liqPay'
}
