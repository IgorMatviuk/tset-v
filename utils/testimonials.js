export default function testimonials(context) {
  return {
    create() {}
  }
}
