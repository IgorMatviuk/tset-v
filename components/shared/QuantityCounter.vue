<template>
  <div class="counter">
    <button
      class="counter__toggle counter__toggle--decrease"
      type="button"
      @click="decreaseValue"
    >
      <svg-icon name="minus" class="counter__icon" height="1" width="17" />
    </button>
    <input v-model="counterValue" class="counter__input" type="text" />
    <button
      class="counter__toggle counter__toggle--increase"
      type="button"
      @click="increaseValue"
    >
      <svg-icon name="plus" class="counter__icon" height="17" width="17" />
    </button>
  </div>
</template>

<script>
import _debounce from 'lodash.debounce'

import { PRODUCT_MAX_QTY } from '@/utils/constants'

export default {
  props: {
    count: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      counterValue: null
    }
  },
  watch: {
    count: {
      handler(newValue, oldValue) {
        /* istanbul ignore else */
        if (newValue !== oldValue) {
          this.counterValue = newValue
        }
      },
      immediate: true
    },
    counterValue: _debounce(function(newValue, oldValue) {
      const isEqual = newValue === oldValue
      const isEdges = newValue === PRODUCT_MAX_QTY || newValue === 1

      if (!isEqual && !isEdges) {
        const parsedValue = parseInt(newValue)
        const isNumber = !isNaN(parsedValue)

        let normilizedValue

        if (isNumber) {
          if (parsedValue > PRODUCT_MAX_QTY) {
            normilizedValue = PRODUCT_MAX_QTY
          } else if (parsedValue > 0) {
            normilizedValue = parsedValue
          } else {
            normilizedValue = 1
          }
        } else {
          normilizedValue = 1
        }

        this.counterValue = normilizedValue
        this.onCounterValueUpdate()
      }
    }, 300)
  },
  methods: {
    decreaseValue() {
      if (this.counterValue > 1) {
        this.counterValue = Number(this.counterValue) - 1
      }
    },
    increaseValue() {
      this.counterValue = Number(this.counterValue) + 1
    },
    onCounterValueUpdate() {
      this.$emit('change', this.counterValue)
    }
  }
}
</script>

<style lang="scss">
.counter {
  display: flex;

  &__input {
    width: 52px;
    height: 27px;
    padding-right: 5px;
    padding-left: 5px;
    font-size: 14px;
    font-weight: 500;
    color: $black;
    text-align: center;
    border: 1px solid $gray;
    border-radius: 4px;
    transition: border-color 0.3s;

    &:focus,
    &:active {
      border-color: $primary;
      outline: none;
      box-shadow: none;
    }
  }

  &__toggle {
    @include reset-button-styles();

    display: flex;
    align-items: center;
    justify-content: center;
    width: 27px;
    height: 27px;

    &--decrease {
      justify-content: flex-start;
    }

    &--increase {
      justify-content: flex-end;
    }
  }

  &__toggle:hover &__icon,
  &__toggle:focus &__icon,
  &__toggle:active &__icon {
    color: $primary;
  }

  &__icon {
    width: 17px;
    height: 17px;
    color: $text-color;
    transition: color 0.3s;
  }

  @include media-breakpoint-down(sm) {
    &__toggle {
      width: 22px;
      height: 22px;
    }

    &__input {
      width: 42px;
      height: 22px;
    }
  }
}
</style>
