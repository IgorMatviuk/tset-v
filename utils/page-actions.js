import { CATALOG_ITEMS_PER_PAGE } from '@/utils/constants'

import { mutationTypes as catalogMutationTypes } from '@/store/catalog'
import { mutationTypes as productMutationTypes } from '@/store/product'
import { actionTypes as searchActionTypes } from '@/store/search'
import { actionTypes as testimonialsActionTypes } from '@/store/testimonials'
import { actionTypes as cartActionTypes } from '@/store/cart'
import { actionsTypes as blogActionsTypes } from '@/store/blog'

function catalog({ app, store, route, page }) {
  const { category, products, attributes } = page.data
  const { meta } = page

  store.commit(`catalog/${catalogMutationTypes.SET_CATALOG_DATA}`, {
    category,
    products,
    attributes
  })

  const filters = store.state.catalog.filters

  if (!filters.length) {
    const localeRegExp = new RegExp(`^/${store.state.locale}/`)
    let filtersPath

    if (category) {
      const categoryIndex = route.path.lastIndexOf(category.slug)
      filtersPath = route.path.slice(categoryIndex + category.slug.length + 1)
    } else {
      const path = route.path.replace(localeRegExp, '/')
      filtersPath = path.replace(/^\/([^/]+)\/?/, '')
    }

    let initialFilters = []

    if (filtersPath) {
      const segments = filtersPath.split('/')

      initialFilters = segments.map((segment) => {
        const [code, ...value] = segment.split('_')
        const label = app.i18n.t(`attributes.${code}.one`)

        return { code, label, value }
      })
    }

    store.commit(`catalog/${catalogMutationTypes.SET_FILTERS}`, initialFilters)

    const totalPages = Math.ceil(meta.total / CATALOG_ITEMS_PER_PAGE)

    store.commit(`catalog/${catalogMutationTypes.SET_TOTAL_PAGES}`, totalPages)

    store.commit(
      `catalog/${catalogMutationTypes.SET_CURRENT_PAGE}`,
      meta.current_page
    )

    store.commit(
      `catalog/${catalogMutationTypes.SET_SEO_TEXT}`,
      meta.tags.seo_text
    )

    const queryOrder = route.query.order

    if (queryOrder) {
      store.commit(`catalog/${catalogMutationTypes.SET_ORDER}`, queryOrder)
    }
  }
}

function product({ store, route, page }) {
  /* eslint-disable camelcase */
  const {
    product,
    links,
    similar_products,
    featured_products,
    viewed_products
  } = page.data

  const sections = {
    similar_products,
    featured_products,
    viewed_products
  }

  const variant = product.variants.find(
    (variant) => variant.url_key === route.params.pathMatch
  )

  store.commit(`product/${productMutationTypes.SET_SELECTED_SIZE}`, null)
  store.commit(`product/${productMutationTypes.SET_PRODUCT_DATA}`, product)
  store.commit(`product/${productMutationTypes.SET_ACTIVE_VARIANT}`, variant)
  store.commit(`product/${productMutationTypes.SET_SECTIONS_DATA}`, sections)
  store.commit(`product/${productMutationTypes.SET_NAV_LINKS}`, links)
}

async function search({ store, route }) {
  await store.dispatch(`search/${searchActionTypes.GET_PRODUCTS}`, {
    query: route.query.q,
    page: route.query.page
  })
}

async function testimonials({ store, route }) {
  const query = {}

  // TODO: remove underscores (mock solution)
  Object.keys(route.query).forEach((key) => {
    if (key !== 'q') {
      query[`_${key}`] = route.query[key]
    } else {
      query[key] = route.query[key]
    }
  })

  await store.dispatch(
    `testimonials/${testimonialsActionTypes.GET_TESTIMONIALS}`,
    query
  )
}

async function blog({ store, route }) {
  const query = {}

  Object.keys(route.query).forEach((key) => {
    query[key] = route.query[key]
  })

  await store.dispatch(`blog/${blogActionsTypes.GET_BLOG_POSTS}`, query)
}

async function checkout({ store }) {
  await store.dispatch(`cart/${cartActionTypes.GET_CHECKOUT_DATA}`)
}

export default { catalog, product, search, checkout, testimonials, blog }
