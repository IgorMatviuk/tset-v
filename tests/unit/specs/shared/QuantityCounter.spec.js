import { shallowMount } from '@vue/test-utils'
import _debounce from 'lodash.debounce'

import QuantityCounter from '@/components/shared/QuantityCounter'

import { PRODUCT_MAX_QTY } from '@/utils/constants'

jest.mock('lodash.debounce', () => jest.fn((fn) => fn))

describe('QuantityCounter', () => {
  beforeEach(() => {
    _debounce.mockClear()
  })

  function createQuantityCounter(initialCount = 1) {
    const wrapper = shallowMount(QuantityCounter, {
      propsData: { count: initialCount },
      stubs: ['svg-icon']
    })

    return wrapper
  }

  describe('Structure', () => {
    test('has expected structure', () => {
      const wrapper = createQuantityCounter()

      expect(wrapper.element).toMatchSnapshot()
      expect(wrapper.vm.count).toMatchSnapshot()

      wrapper.setProps({ count: wrapper.vm.count + 1 })

      expect(wrapper.element).toMatchSnapshot()
      expect(wrapper.vm.count).toMatchSnapshot()
    })
  })

  describe('Change quantity', () => {
    test('increase quantity', async () => {
      const wrapper = createQuantityCounter(2)
      const btn = wrapper.find('.counter__toggle--decrease')

      expect(wrapper.vm.counterValue).toBe(2)
      await btn.trigger('click')
      expect(wrapper.vm.counterValue).toBe(1)

      await btn.trigger('click')
      expect(wrapper.vm.counterValue).toBe(1)
    })

    test('decrease quantity', async () => {
      const wrapper = createQuantityCounter(1)
      const btn = wrapper.find('.counter__toggle--increase')

      expect(wrapper.vm.counterValue).toBe(1)
      await btn.trigger('click')
      expect(wrapper.vm.counterValue).toBe(2)

      await btn.trigger('click')
      expect(wrapper.vm.counterValue).toBe(3)
    })

    test('manually change quantity', async () => {
      const wrapper = createQuantityCounter()
      const input = wrapper.find('.counter__input')

      // Set a value grater than a max
      await input.setValue(PRODUCT_MAX_QTY + 1)
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.counterValue).toBe(PRODUCT_MAX_QTY)
      expect(wrapper.emitted('change').length).toBe(1)

      // Set a string value
      await input.setValue('invalid')
      expect(wrapper.vm.counterValue).toBe(1)

      // Set a negative value
      await input.setValue(-1)
      expect(wrapper.vm.counterValue).toBe(1)
    })
  })
})
