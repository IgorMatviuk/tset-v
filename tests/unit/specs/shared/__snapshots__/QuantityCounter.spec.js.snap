// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityCounter Structure has expected structure 1`] = `
<div
  class="counter"
>
  <button
    class="counter__toggle counter__toggle--decrease"
    type="button"
  >
    <svg-icon-stub
      class="counter__icon"
      height="1"
      name="minus"
      width="17"
    />
  </button>
   
  <input
    class="counter__input"
    type="text"
  />
   
  <button
    class="counter__toggle counter__toggle--increase"
    type="button"
  >
    <svg-icon-stub
      class="counter__icon"
      height="17"
      name="plus"
      width="17"
    />
  </button>
</div>
`;

exports[`QuantityCounter Structure has expected structure 2`] = `1`;

exports[`QuantityCounter Structure has expected structure 3`] = `
<div
  class="counter"
>
  <button
    class="counter__toggle counter__toggle--decrease"
    type="button"
  >
    <svg-icon-stub
      class="counter__icon"
      height="1"
      name="minus"
      width="17"
    />
  </button>
   
  <input
    class="counter__input"
    type="text"
  />
   
  <button
    class="counter__toggle counter__toggle--increase"
    type="button"
  >
    <svg-icon-stub
      class="counter__icon"
      height="17"
      name="plus"
      width="17"
    />
  </button>
</div>
`;

exports[`QuantityCounter Structure has expected structure 4`] = `2`;
